src/customElements.ts

```diff
@@ -0,0 +1,198 @@
+	let { propertyName = attributeName, value = attributeValue } = descriptor;
+
+	if (typeof value === 'function') {
+		value = value(attributeValue);
+	}
+
+	return [ propertyName, value ];
+}
+
+/**
+ * Called by HTMLElement subclass to initialize itself with the appropriate attributes/properties/events.
+ *
+ * @param element
+ */
+export function initializeElement(element: CustomElement) {
+	let initialProperties: any = {};
+
+	const { attributes = [], events = [], properties = [], initialization } = element.getDescriptor();
+
+	attributes.forEach(attribute => {
+		const attributeName = attribute.attributeName;
+
+		const [ propertyName, propertyValue ] = getWidgetPropertyFromAttribute(attributeName, element.getAttribute(attributeName), attribute);
+		initialProperties[ propertyName ] = propertyValue;
+	});
+
+	let customProperties: PropertyDescriptorMap = {};
+
+	attributes.reduce((properties, attribute) => {
+		const { propertyName = attribute.attributeName } = attribute;
+
+		properties[ propertyName ] = {
+			get() {
+				return element.getWidgetInstance().properties[ propertyName ];
+			},
+			set(value: any) {
+				const [ propertyName, propertyValue ] = getWidgetPropertyFromAttribute(attribute.attributeName, value, attribute);
+				element.getWidgetInstance().setProperties(assign({}, element.getWidgetInstance().properties, {
+					[propertyName]: propertyValue
+				}));
+			}
+		};
+
+		return properties;
+	}, customProperties);
+
+	properties.reduce((properties, property) => {
+		const { propertyName, getValue, setValue } = property;
+		const { widgetPropertyName = propertyName } = property;
+
+		properties[ propertyName ] = {
+			get() {
+				const value = element.getWidgetInstance().properties[ widgetPropertyName ];
+				return getValue ? getValue(value) : value;
+			},
+
+			set(value: any) {
+				element.getWidgetInstance().setProperties(assign(
+					{},
+					element.getWidgetInstance().properties,
+					{ [widgetPropertyName]: setValue ? setValue(value) : value }
+				));
+			}
+		};
+
+		return properties;
+	}, customProperties);
+
+	Object.defineProperties(element, customProperties);
+
+	// define events
+	events.forEach((event) => {
+		const { propertyName, eventName } = event;
+
+		initialProperties[ propertyName ] = (event: any) => {
+			element.dispatchEvent(new CustomEvent(eventName, {
+				bubbles: false,
+				detail: event
+			}));
+		};
+	});
+
+	// find children
+	let children: DNode[] = [];
+
+	Array.prototype.slice.call(element.children, 0).forEach((childNode: HTMLElement, index: number) => {
+		children.push(w(createDomWrapper, {
+			key: `child-${index}`,
+			domNode: childNode
+		}));
+	});
+
+	if (initialization) {
+		initialization.call(element, initialProperties);
+	}
+
+	Array.prototype.slice.call(element.children, 0).forEach((childNode: HTMLElement) => {
+		element.removeChild(childNode);
+	});
+
+	const widgetInstance = element.getWidgetFactory().mixin(createProjectorMixin)({
```

Comment:
Can you tell it here that it is now a Projector? and avoid the cast to `any` below?